### PR TITLE
Add scan function with pagination

### PR DESCRIPTION
**Pull request by tkyi**
This pull request adds the `scan` function
It does not account for any query parameters/filtering

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -157,6 +157,65 @@ datastore.update({
...
+
+**Arguments**
+
+* `config` - An `object`. Each of its properties defines your save operation
+* `config.table` - A `string`. The DynamoDB table name
+* `config.params` - An `object`. Each of its properties defines the query parameters
+* `config.paginate` - An `object`. Each of its properties defines the pagination parameters
+* `config.paginate.count` - A `number`. The number of items per page
+* `config.paginate.page` - An `number`. This is the page number of the set you wish for the datastore to return
+* `callback(err, result)`  - A callback which is called when the task is completed. It receives the `err` and `result`, where `result` is the data that was saved in the datastore. Returns `null` if the record is empty. Returns error if the record does
```
Comment (d2lam) [inline README.md]: `result` is the list of records. Returns `[]` if no records are found.

File: index.js [hunk 2]
```diff
@@ -143,8 +143,40 @@ class Dynamodb extends Datastore {
...
+     * @method scan
+     * @param  {Object}   config                Configuration object
+     * @param  {String}   config.table          Table name
+     * @param  {Object}   [config.params]       Record data
+     * @param  {Object}   config.paginate       Pagination parameters
+     * @param  {Number}   config.paginate.count Number of items per page
+     * @param  {Number}   config.paginate.page  Specific page of the set to return
+     * @param  {Function} callback              fn(err, data)
+     *                                          err - Error object
+     *                                          data - Data saved in the table
```
Comment (d2lam) [inline index.js]: data is the list of records found

File: index.js [hunk 3]
```diff
@@ -143,8 +143,40 @@ class Dynamodb extends Datastore {
...
+        const endIndex = config.paginate.page * config.paginate.count;
+        const startIndex = (config.paginate.page - 1) * config.paginate.count;
+
+        if (!client) {
+            const err = new Error(`Invalid table name "${config.table}"`);
+
+            return callback(err);
+        }
+
+        return client.scan().limit(endIndex)  // scan all items up to and including last item specified
```
Comment (d2lam) [inline index.js]: If `client.scan` includes the last item then it should be `(endIndex-1)` here

For example, the records are `[a,b,c,d,e,f]`
`page:2 count:2` should give `c,d`
Start index = `(2-1)*2 = 2`
End index = `2*2 = 4`

File: test/index.test.js [hunk 4]
```diff
@@ -22,11 +23,18 @@ describe('index test', () => {
 
     beforeEach(() => {
         responseMock = {
-            toJSON: sinon.stub()
+            toJSON: sinon.stub(),
+            Items: []
+        };
+        scanChainMock = {
+            limit: sinon.stub(),
+            loadAll: sinon.stub(),
```
Comment (d2lam) [inline test/index.test.js]: Why do we need this?

File: test/index.test.js [hunk 5]
```diff
@@ -381,4 +389,113 @@ describe('index test', () => {
...
+                    key: 'value'
+                },
+                {
+                    id: 'data',
+                    key: 'value'
+                }
+            ];
+
+            scanChainMock.limit.returns(scanChainMock);
+            scanChainMock.loadAll.returns(scanChainMock);
```
Comment (d2lam) [inline test/index.test.js]: Do we use this for something?

**Comment by d2lam** [comment]
Going to need to update this using the new data schema since this is using version 4 right now. But that can be in a separate PR.

File: test/index.test.js (hunk 4 above)
Comment (tkyi) [inline test/index.test.js]: It was part of the scan() chain before but I realized loadAll() actually loads all entries in the table https://github.com/screwdriver-cd/datastore-dynamodb/pull/8/files#diff-168726dbe96b3ce427e7fedce31bb0bcR170

File: test/index.test.js (hunk 5 above)
Comment (tkyi) [inline test/index.test.js]: No, I forgot to delete it after removing loadAll() from the code
Good catch!

File: README.md [hunk 6]
```diff
@@ -157,6 +157,64 @@ datastore.update({
...
+
+**Arguments**
+
+* `config` - An `object`. Each of its properties defines your save operation
+* `config.table` - A `string`. The DynamoDB table name
+* `config.params` - An `object`. Each of its properties defines the query parameters
+* `config.paginate` - An `object`. Each of its properties defines the pagination parameters
+* `config.paginate.count` - A `number`. The number of items per page
+* `config.paginate.page` - An `number`. This is the page number of the set you wish for the datastore to return
+* `callback(err, result)`  - A callback which is called when the task is completed. It receives the `err` and `result`, where `result` is the list of records in the datastore. Returns `[]` if the record is empty. Returns error if the record does not exist.
```
Comment (d2lam) [inline README.md]: `Returns error if the record does not exist`?
What does that mean? It returns `[]` if there is nothing in the datastore right?

File: index.js [hunk 7]
```diff
@@ -143,8 +143,40 @@ class Dynamodb extends Datastore {
...
+     *                                          data - List of records in the table
+     */
     scan(config, callback) {
-        callback();
+        const client = this.client[config.table];
+        const endIndex = config.paginate.page * config.paginate.count;
+        const startIndex = (config.paginate.page - 1) * config.paginate.count;
+
+        if (!client) {
+            const err = new Error(`Invalid table name "${config.table}"`);
```
Comment (d2lam) [inline index.js:165]: Shouldn't this return `[]` if table does not exist?

File: README.md (hunk 6 above)
Comment (tkyi) [inline README.md]: Mm trying to say returns error if `tableName` does not exist

File: index.js (hunk 7 above)
Comment (d2lam) [inline index.js:165]: https://github.com/screwdriver-cd/screwdriver-datastore-imdb/blob/master/lib/imdb.js#L47-L49

Comment (tkyi) [inline index.js:165]: We will keep the error message for now and work to make the two datastores consistent in another pull request

https://github.com/screwdriver-cd/datastore-dynamodb/issues/9

File: README.md [hunk 8]
```diff
@@ -157,6 +157,64 @@ datastore.update({
...
+
+**Example**
+
+```js
+const DynamoDB = require('screwdriver-datastore-imdb');
```
Comment (nkatzman) [inline README.md]: should be `screwdriver-datastore-dynamodb`

File: README.md [hunk 9]
```diff
@@ -157,6 +157,64 @@ datastore.update({
 });
 ```
 
+###  scan
+
+Scan a record in the datastore. Returns `[]` if the record is empty. Returns error if the `tableName` does not exist.
```
Comment (nkatzman) [inline README.md]: `Scan a table in the datastore. Return [] if the table is empty. Returns error if the table does not exist`

File: index.js [hunk 10]
```diff
@@ -143,8 +143,40 @@ class Dynamodb extends Datastore {
...
+
+        return client.scan().limit(endIndex - 1)  // scan all items up to and including last item specified
+        .exec((err, data) => {
+            if (err) {
+                return callback(err);
+            }
+            const result = data.Items.slice(startIndex);    // pick out items from page specified
+            const response = result.map((item) => item.toJSON());
+
+            return callback(err, response);
```
Comment (nkatzman) [inline index.js]: You don't need to be passing back `err` here, even though it will be a null value.

Instead pass `callback(null, response)` so that it's more explicit

File: package.json [hunk 11]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "screwdriver-datastore-dynamodb",
-  "version": "0.0.1",
+  "version": "1.0.0",
   "description": "hapi interface with dynamodb",
```
Comment (nkatzman) [inline package.json]: this is a screwdriver inferface with dynamodb

**Comment by nkatzman** [comment]
Let's do it 👍

**Comment by d2lam** [comment]
👍
